File: src/main/java/io/github/oxnz/Ingrid/UserRepository.java
package io.github.oxnz.Ingrid;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.oxnz.Ingrid.user.User;
import org.springframework.data.redis.core.HashOperations;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Repository;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

@Repository
public class UserRepository {

    private static final String KEY = "user";
    private ObjectMapper objectMapper;
    private StringRedisTemplate redisTemplate;
    private HashOperations hashOperations;

    public UserRepository(ObjectMapper objectMapper, StringRedisTemplate redisTemplate) {
        this.objectMapper = objectMapper;
        this.redisTemplate = redisTemplate;
        hashOperations = redisTemplate.opsForHash();
    }

    private User readUser(String s) {
        try {
            return objectMapper.readValue(s, User.class);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private String writeUser(User user) {
        try {
            return objectMapper.writeValueAsString(user);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    public User get(long id) {
        String s = (String) hashOperations.get(KEY, String.valueOf(id));
        return readUser(s);
    }

    public List<User> list() {
        List<String> ss = hashOperations.values(KEY);
        return ss
                .stream()
                .map(this::readUser)
                .collect(Collectors.toList());
    }

    public void put(User user) {
        String s = writeUser(user);
        hashOperations.put(KEY, String.valueOf(user.id()), s);
    }

}

File: src/main/scala/io/github/oxnz/Ingrid/user/UserRepository.scala
package io.github.oxnz.Ingrid.user

import java.io.IOException

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.data.redis.core.{HashOperations, StringRedisTemplate}
import org.springframework.stereotype.Repository

import scala.jdk.CollectionConverters._

object UserRepository {
  private val KEY = "user"
}

@Repository class UserRepository(var objectMapper: ObjectMapper, var redisTemplate: StringRedisTemplate) {
  private var hashOperations: HashOperations[String, String, String] = redisTemplate.opsForHash()

  private def readUser(s: String) = try objectMapper.readValue(s, classOf[User])
  catch {
    case e: IOException =>
      throw new RuntimeException(e)
  }

  private def writeUser(user: User) = try objectMapper.writeValueAsString(user)
  catch {
    case e: JsonProcessingException =>
      throw new RuntimeException(e)
  }

  def findOne(id: Long): User = {
    val s = hashOperations.get(UserRepository.KEY, String.valueOf(id)).asInstanceOf[String]
    readUser(s)
  }

  def findAll: List[User] = {
    val ss = hashOperations.values(UserRepository.KEY).asScala.toList
    ss.map(readUser)
  }

  def save(user: User): Unit = {
    val s = writeUser(user)
    hashOperations.put(UserRepository.KEY, String.valueOf(user.id), s)
  }
}